server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -131,27 +121,46 @@ public RemoteStoreRefreshListener(
             @Override
             public void beforeUpload(String file) {
                 // Start tracking the upload bytes started
-                segmentTracker.addUploadBytesStarted(latestFileNameSizeOnLocalMap.get(file));
+                segmentTracker.addUploadBytesStarted(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
             }
 
             @Override
             public void onSuccess(String file) {
                 // Track upload success
-                segmentTracker.addUploadBytesSucceeded(latestFileNameSizeOnLocalMap.get(file));
+                segmentTracker.addUploadBytesSucceeded(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
                 segmentTracker.addToLatestUploadedFiles(file);
             }
 
             @Override
             public void onFailure(String file) {
                 // Track upload failure
-                segmentTracker.addUploadBytesFailed(latestFileNameSizeOnLocalMap.get(file));
+                segmentTracker.addUploadBytesFailed(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
             }
         };
     }
 
     @Override
     public void beforeRefresh() throws IOException {}
 
+    @Override
+    protected void runAfterRefreshExactlyOnce(boolean didRefresh) {
+        if (shouldSync(didRefresh)) {
+            updateLocalRefreshTimeAndSeqNo();
+            try {
+                if (this.primaryTerm != indexShard.getOperationPrimaryTerm()) {
+                    this.primaryTerm = indexShard.getOperationPrimaryTerm();
+                    this.remoteDirectory.init();
+                }
+                try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
+                    Collection<String> localSegmentsPostRefresh = segmentInfosGatedCloseable.get().files(true);
+                    updateLocalSizeMapAndTracker(localSegmentsPostRefresh);
+                }
+            } catch (Throwable t) {
+                logger.error("Exception in runAfterRefreshExactlyOnce() method", t);
```

Comment:
It is possible that `runAfterRefreshExactlyOnce` fails but ` runAfterRefreshWithPermit` succeeds. This is not going to create any inconsistent state because the `updateLocalSizeMapAndTracker` method will be invoked again in the `runAfterRefreshWithPermit`.

Reply:
Got it. If it is getting invoked in `runAfterRefreshWithPermit`, do we need to invoke it here as well?

Reply:
As discussed, summarising the 2 cases -
 -  If there is a refresh that got triggered, but a retry is already going on (and is pass the `updateLocalSizeMapAndTracker` method) and takes couple of minutes. In such cases if we keep the `updateLocalSizeMapAndTracker` only within `syncSegments` method, then the bytes lag would be inconsistent for as long as the upload takes.
 - If a refresh happened, and before the afterRefresh of the RemoteStoreRefreshListener could be triggered, a retry kicked in. In such cases, the upload bytes would be calculated wrongly and shown as is in the response of remote store stats api.